crates/icrate/examples/metal.rs

```diff
@@ -0,0 +1,344 @@
+                    .cast::<core::ffi::c_void>();
+            unsafe {
+                encoder.setVertexBytes_length_atIndex(
+                    scene_properties_bytes,
+                    core::mem::size_of::<SceneProperties>(),
+                    0,
+                )
+            };
+
+            let vertex_input_data: &mut [VertexInput] = &mut [
+                VertexInput {
+                    position: Position {
+                        x: -f32::sqrt(3.0) / 4.0,
+                        y: -0.25,
+                        z: 0.,
+                    },
+                    color: Color {
+                        r: 1.,
+                        g: 0.,
+                        b: 0.,
+                    },
+                },
+                VertexInput {
+                    position: Position {
+                        x: f32::sqrt(3.0) / 4.0,
+                        y: -0.25,
+                        z: 0.,
+                    },
+                    color: Color {
+                        r: 0.,
+                        g: 1.,
+                        b: 0.,
+                    },
+                },
+                VertexInput {
+                    position: Position {
+                        x: 0.,
+                        y: 0.5,
+                        z: 0.,
+                    },
+                    color: Color {
+                        r: 0.,
+                        g: 0.,
+                        b: 1.,
+                    },
+                },
+            ];
+            let vertex_input_bytes =
+                unsafe { NonNull::new_unchecked(vertex_input_data) }.cast::<core::ffi::c_void>();
+            unsafe {
+                encoder.setVertexBytes_length_atIndex(
+                    vertex_input_bytes,
+                    vertex_input_data.len() * core::mem::size_of::<VertexInput>(),
+                    1,
+                )
+            };
+
+            encoder.setRenderPipelineState(&self.pipeline_state);
+            unsafe {
+                encoder.drawPrimitives_vertexStart_vertexCount(MTLPrimitiveTypeTriangle, 0, 3)
+            };
+            encoder.endEncoding();
+            command_buffer.presentDrawable(&current_drawable);
+            command_buffer.commit();
+        }
+
+        #[method(mtkView:drawableSizeWillChange:)]
+        #[allow(non_snake_case)]
+        unsafe fn mtkView_drawableSizeWillChange(&self, _view: &MTKView, _size: NSSize) {
+            // println!("mtkView_drawableSizeWillChange");
+        }
+    }
+);
+
+unsafe impl NSObjectProtocol for Delegate {}
+
+#[cfg(target_os = "macos")]
+impl Delegate {
+    pub fn init_with_shaders(shaders: &NSString) -> Id<Self> {
+        unsafe { msg_send_id![Self::alloc(), initWithShaders: shaders] }
+    }
+}
+
+fn main() {
+    let app = unsafe { NSApplication::sharedApplication() };
+    unsafe { app.setActivationPolicy(NSApplicationActivationPolicyRegular) };
+
+    #[rustfmt::skip]
+    let shaders = ns_string!(r#"
```

Comment:
Nit: Move the shader to a constant at the top of the file, so that it's easier to see what the Rust structs we define are used for.